### PR TITLE
Undo support for multiple parented items in outliner

### PR DESCRIPTION
**Pull request by HamedSabri-adsk**
Summary of changes:

* Adopt to new UFE changes: (**Ufe::Scene**, **Ufe::SceneCompositeNotfication**, **Ufe::hierarchy**)
* Translate/Rotate/Scale/RotatePivotTranslate commands now rely on Ufe::Path.
* Added a new testObservableScene.py

### CODE REVIEW

File: test/lib/ufe/testObservableScene.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+        ufe.Scene.notify(ufe.ObjectAdd(itemA))
+
+        # we should now have an ObjectAdd notification
+        self.checkNotifications(snObs, [1,0,0,0,0,0])
+
+        # HS 2020: can't pass ufe.scene to NotificationGuard??
+        
+        # Composite notifications with guard.
+        # with ufe.NotificationGuard(ufe.Scene):
```
Comment (HamedSabri-adsk) [inline test/lib/ufe/testObservableScene.py:100]: @ppt-adsk there was a pybind11 error when I tried to pass ufe,Scene to NotificationGuard. I will complete the remaining of the test once this issue is fixed.

File: lib/mayaUsd/ufe/UsdSceneItem.cpp [hunk 2]
```diff
@@ -53,7 +53,13 @@ std::string UsdSceneItem::nodeType() const
...
 {
 	std::vector<std::string> strAncestorTypes;
-
+	
+	// during drag/drop in outliner UsdPrim can become invalidated which
+	// should never happens.
+	// MAYA-106644: HS August 31,2020 investigate the invalidation.
+	if (!fPrim.IsValid()) {
+		return {};
+	}
```
Comment (HamedSabri-adsk) [inline lib/mayaUsd/ufe/UsdSceneItem.cpp:62]: @seando-adsk I added this temporarily workaround to pass the unexpected crash with my recent changes.

File: lib/mayaUsd/ufe/StagesSubject.cpp [hunk 3]
```diff
@@ -223,13 +223,11 @@ void StagesSubject::stageChanged(UsdNotice::ObjectsChanged const& notice, UsdSta
 			{
 				if (prim.IsActive())
 				{
-					auto notification = Ufe::ObjectAdd(sceneItem);
-					Ufe::Scene::notifyObjectAdd(notification);
```
Comment (seando-adsk) [inline lib/mayaUsd/ufe/StagesSubject.cpp:227]: I believe that all your changes will need to keep this old code around and protect your new code with UFE preview version check. Something like this:

```
#if UFE_PREVIEW_VERSION_NUM >= 2021
    <insert new code here>
#else
    <keep old code here>
#endif
```

File: test/lib/ufe/testObservableScene.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+class UFEObservableSceneTest(unittest.TestCase):
+
+    def checkNotifications(self, testObserver, listNotifications):
+        self.assertTrue(testObserver.add == listNotifications[0])
+        self.assertTrue(testObserver.delete == listNotifications[1])
+        self.assertTrue(testObserver.pathChange == listNotifications[2])
+        self.assertTrue(testObserver.subtreeInvalidate == listNotifications[3])
+        self.assertTrue(testObserver.composite == listNotifications[4])
+
+    def testObservableSelection(self):
```
Comment (ppt-adsk) [inline test/lib/ufe/testObservableScene.py:59]: Copy-paste damage: should be testObservableScene(), or testSceneObservation(), or similar.

File: lib/mayaUsd/ufe/UsdSceneItem.cpp (hunk 2 above)
Comment (seando-adsk) [reply lib/mayaUsd/ufe/UsdSceneItem.cpp:62]: This was a bug that I recently fixed with commit 4fbdf7f2a4f12fae6bcaef634faff3b3c626d198. 

I would argue that we should also not have checks like this. If the prim is invalid, then this entire UsdSceneItem is invalid and something has gone wrong. It should crash (or at least assert) so we can find and fix the real problem.

File: test/lib/ufe/testObservableScene.py (hunk 1 above)
Comment (ppt-adsk) [reply test/lib/ufe/testObservableScene.py:100]: Let me know if you need some help there.

File: test/lib/ufe/testObservableScene.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            self.pathChange += 1
+        if isinstance(notification, ufe.SubtreeInvalidate):
+            self.subtreeInvalidate += 1
+        if isinstance(notification, ufe.SceneCompositeNotification):
+            self.composite += 1
+
+    def notifications(self):
+        return [self.add, self.delete, self.pathChange, self.subtreeInvalidate, self.composite]
+
+class UFEObservableSceneTest(unittest.TestCase):
```
Comment (seando-adsk) [inline test/lib/ufe/testObservableScene.py:50]: You need a skipIf tag here so this test only runs if UFE v >= 2021

File: test/lib/ufe/testRename.py [hunk 6]
```diff
@@ -420,14 +420,10 @@ def testRenameNotifications(self):
...
 
         # We start off with no observers
-        self.assertFalse(ufe.Scene.hasObjectAddObserver(ufeObs))
-        self.assertFalse(ufe.Scene.hasObjectDeleteObserver(ufeObs))
-        self.assertFalse(ufe.Scene.hasObjectPathChangeObserver(ufeObs))
-
-        # Add the UFE observers we want to test
-        ufe.Scene.addObjectAddObserver(ufeObs)
-        ufe.Scene.addObjectDeleteObserver(ufeObs)
-        ufe.Scene.addObjectPathChangeObserver(ufeObs)
```
Comment (seando-adsk) [inline test/lib/ufe/testRename.py:430]: Again you need to keep the old test code around along with your new code and use an if for UFE preview version.

File: lib/mayaUsd/ufe/UsdSceneItem.cpp (hunk 2 above)
Comment (HamedSabri-adsk) [reply lib/mayaUsd/ufe/UsdSceneItem.cpp:62]: @seando-adsk  Yes, this was just a temporarily check so I can investigate the issue with the new recent changes both in Maya and maya-usd. This block of code will disappear soon today.  Just ignore it.

I am looking into this as we speak and the problem is that in the middle of composite notification, your newly added **iconAt()** is getting called which shouldn't happen. We can talk internally to show you how this happens.

Comment (HamedSabri-adsk) [reply lib/mayaUsd/ufe/UsdSceneItem.cpp]: addressed in 3fa7e6a9d95276e1fb7ca02d28cac92ca9995f59

File: lib/mayaUsd/ufe/UsdTRSUndoableCommandBase.h [hunk 7]
```diff
@@ -80,30 +78,31 @@ class MAYAUSD_CORE_PUBLIC UsdTRSUndoableCommandBase : public Ufe::Observer,
     virtual void addEmptyAttribute() = 0;
     virtual bool cannotInit() const;
 
-private:
-
-    // Overridden from Ufe::Observer
-    void operator()(const Ufe::Notification& notification) override;
+    // Create a UsdSceneItem::Ptr conditionaly in the first access from Ufe::Path.
```
[the remark this answers is not in the document]
Comment (HamedSabri-adsk) [reply lib/mayaUsd/ufe/UsdTRSUndoableCommandBase.h]: addressed in : 871a903669aeceb3ac75c1d2e4d515e1a0144a69

File: lib/mayaUsd/ufe/UsdSceneItem.cpp [hunk 8]
```diff
@@ -53,7 +53,13 @@ std::string UsdSceneItem::nodeType() const
 std::vector<std::string> UsdSceneItem::ancestorNodeTypes() const
 {
 	std::vector<std::string> strAncestorTypes;
-
+	
+	// during drag/drop in outliner UsdPrim can become invalidated which
```
[the remark this answers is not in the document]
Comment (HamedSabri-adsk) [reply lib/mayaUsd/ufe/UsdSceneItem.cpp]: addressed in 3fa7e6a9d95276e1fb7ca02d28cac92ca9995f59

File: test/lib/ufe/testObservableScene.py (hunk 4 above)
Comment (HamedSabri-adsk) [reply test/lib/ufe/testObservableScene.py]: addressed in: 871a903669aeceb3ac75c1d2e4d515e1a0144a69

File: lib/mayaUsd/ufe/StagesSubject.cpp (hunk 3 above)
Comment (HamedSabri-adsk) [reply lib/mayaUsd/ufe/StagesSubject.cpp:227]: addressed f4f6bd4a848585dfb332e3ee75ac98923af250b5

File: test/lib/ufe/testObservableScene.py (hunk 5 above)
Comment (HamedSabri-adsk) [reply test/lib/ufe/testObservableScene.py:51]: addressed f4f6bd4a848585dfb332e3ee75ac98923af250b5

File: test/lib/ufe/testRename.py (hunk 6 above)
Comment (HamedSabri-adsk) [reply test/lib/ufe/testRename.py:430]: addressed f4f6bd4a848585dfb332e3ee75ac98923af250b5

Also fixed some additional unit tests that were failing:  **testContextOps** , **testDeleteCmd**

File: lib/mayaUsd/ufe/ProxyShapeHierarchy.cpp [hunk 9]
```diff
@@ -158,10 +158,18 @@ Ufe::AppendedChild ProxyShapeHierarchy::appendChild(const Ufe::SceneItem::Ptr& c
 #endif
 
 #ifdef UFE_V2_FEATURES_AVAILABLE
+
+#if UFE_PREVIEW_VERSION_NUM >= 2021
```
Comment (ppt-adsk) [inline lib/mayaUsd/ufe/ProxyShapeHierarchy.cpp:162]: Nit-pick: I would have ifdef'ed only the first line, but it doesn't really matter.

File: lib/mayaUsd/ufe/UsdHierarchy.h [hunk 10]
```diff
@@ -62,7 +63,7 @@ class MAYAUSD_CORE_PUBLIC UsdHierarchy : public Ufe::Hierarchy
 #endif
 
 #ifdef UFE_V2_FEATURES_AVAILABLE
-    Ufe::UndoableCommand::Ptr insertChildCmd(
+    Ufe::InsertChildCommand::Ptr insertChildCmd(
```
Comment (seando-adsk) [inline lib/mayaUsd/ufe/UsdHierarchy.h:66]: This needs a UFE preview check with new and old code.

File: lib/mayaUsd/ufe/UsdUndoInsertChildCommand.h [hunk 11]
```diff
@@ -28,7 +28,7 @@ MAYAUSD_NS_DEF {
 namespace ufe {
 
 //! \brief UsdUndoInsertChildCommand
-class MAYAUSD_CORE_PUBLIC UsdUndoInsertChildCommand : public Ufe::UndoableCommand
+class MAYAUSD_CORE_PUBLIC UsdUndoInsertChildCommand : public Ufe::InsertChildCommand
```
Comment (seando-adsk) [inline lib/mayaUsd/ufe/UsdUndoInsertChildCommand.h:31]: Doesn't this need a UFE preview check with new and old code?

File: lib/mayaUsd/ufe/Utils.h [hunk 12]
```diff
@@ -85,7 +85,7 @@ template <class T>
 void sendNotification(const Ufe::SceneItem::Ptr& item, const Ufe::Path& previousPath)
 {
     T notification(item, previousPath);
-    Ufe::Scene::notifyObjectPathChange(notification);
+    Ufe::Scene::instance().notify(notification);
```
Comment (seando-adsk) [inline lib/mayaUsd/ufe/Utils.h:88]: Doesn't this need a UFE preview check with new and old code?

File: lib/mayaUsd/render/vp2RenderDelegate/proxyRenderDelegate.cpp [hunk 13]
```diff
@@ -445,7 +445,11 @@ void ProxyRenderDelegate::_InitRenderDelegate(MSubSceneContainer& container) {
                 globalSelection->addObserver(_observer);
             }
 
-            Ufe::Scene::instance().addObjectAddObserver(_observer);
+            #if UFE_PREVIEW_VERSION_NUM >= 2021
+                Ufe::Scene::instance().addObserver(_observer);
+            #else
+                Ufe::Scene::instance().addObjectAddObserver(_observer);
+            #endif
```
Comment (seando-adsk) [inline lib/mayaUsd/render/vp2RenderDelegate/proxyRenderDelegate.cpp:452]: I guess its minor and a matter of style, but why are the #if/else/endif indented? That makes the other two lines indented. So when we remove this ufe preview check you have to unindent the code back.

File: lib/mayaUsd/ufe/ProxyShapeHierarchy.h [hunk 14]
```diff
@@ -66,10 +66,11 @@ class MAYAUSD_CORE_PUBLIC ProxyShapeHierarchy : public Ufe::Hierarchy
...
 #ifdef UFE_V2_FEATURES_AVAILABLE
-    Ufe::UndoableCommand::Ptr insertChildCmd(
-        const Ufe::SceneItem::Ptr& child,
-        const Ufe::SceneItem::Ptr& pos
-    ) override;
+	#if UFE_PREVIEW_VERSION_NUM >= 2021
+    	Ufe::InsertChildCommand::Ptr insertChildCmd(const Ufe::SceneItem::Ptr& child, const Ufe::SceneItem::Ptr& pos) override;
+    #else
+    	Ufe::UndoableCommand::Ptr insertChildCmd(const Ufe::SceneItem::Ptr& child, const Ufe::SceneItem::Ptr& pos) override;
+    #endif
```
Comment (seando-adsk) [inline lib/mayaUsd/ufe/ProxyShapeHierarchy.h:73]: The indentation here weird and things don't line up.

**Comment by HamedSabri-adsk** [comment]
@ppt-adsk  @seando-adsk  See 70da98fde3ed6c4556e6ea63264197192e35b14a

There were many more places where I had to guard against new UFE 20.21 changes. I tried my best to minimize the use of IF/DEF to not clutters things but honestly I am not too crazy to be perfect because this IF/DEFs are going to be cleaned up as soon as the new PR is out and second we are not really following our C++ guideline in lib/ufe which really makes this changes even harder ( mix of spaces/tab, indentation, variable names, etc.... ).

**Review by seando-adsk — APPROVED** [review]

**Review by ppt-adsk — APPROVED** [review]

**Comment by HamedSabri-adsk** [comment]
I will remove the do-not-merge-yet tag once the corresponding MAYA's PR gets merged in first.

**Review by ppt-adsk — APPROVED** [review]

**Review by ppt-adsk — APPROVED** [review]

**Pull request merged**